pkg/log/log_file_test.go

```diff
@@ -0,0 +1,191 @@
+	klogDefaultMaxSize = klog.MaxSize
+)
+
+func initFlags() *pflag.FlagSet {
+	flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
+	AddFlags(flags)
+	flags.AddGoFlagSet(flag.CommandLine)
+	return flags
+}
+
+func restoreFlagDefaultValues() {
+	testFlags.Set(logToStdErrFlag, "true")
+	testFlags.Set(logFileFlag, "")
+	testFlags.Set(logDirFlag, "")
+	testFlags.Set(maxSizeFlag, fmt.Sprintf("%d", klogDefaultMaxSize/1024/1024))
+	testFlags.Set(maxNumFlag, "0")
+
+	klog.MaxSize = klogDefaultMaxSize
+	logFileMaxNum = 0
+	logDir = ""
+}
+
+// Log ~100K bytes.
+func testLogging() {
+	// ~1K bytes per line.
+	line := make([]byte, 1000)
+	for i := 0; i < 1000; i++ {
+		line[i] = byte('0' + i%10)
+	}
+	// Log 100 lines, ~100K bytes.
+	for i := 0; i < 100; i++ {
+		klog.Infof("%d: %s", i, string(line))
+		klog.Warningf("%d: %s", i, string(line))
+	}
+	logs.FlushLogs()
+}
+
+func TestKlogFileLimits(t *testing.T) {
+	testLogDir, err := ioutil.TempDir("", "antrea-log-test")
+	if err != nil {
+		t.Errorf("Failed to create tmp log dir: %v", err)
+		return
+	}
+	defer os.RemoveAll(testLogDir)
+
+	args := []string{"--logtostderr=false", "--log_dir=" + testLogDir, "--log_file_max_size=1", "--log_file_max_num=2"}
+	testFlags.Parse(args)
+	InitLogFileLimits(testFlags)
+	logs.InitLogs()
+
+	// Should generate 3 log files (100K * 30 / 1M).
+	for i := 0; i < 30; i++ {
+		testLogging()
+		time.Sleep(time.Millisecond * 50)
+	}
+
+	infoLogFileNum := 0
+	warningLogFileNum := 0
+	validateFn := func() {
+		f, err := os.Open(testLogDir)
+		if err != nil {
+			t.Errorf("Failed to open log directory: %v", err)
+			return
+		}
+		allFiles, err := f.Readdir(-1)
+		if err != nil {
+			t.Errorf("Failed to read log directory: %v", err)
+			return
+		}
+		f.Close()
+
+		infoLogFiles := []os.FileInfo{}
+		warningLogFiles := []os.FileInfo{}
+		for _, file := range allFiles {
+			if !file.Mode().IsRegular() {
+				// Skip dir, symbol link, etc.
+				continue
+			}
+			if !strings.HasPrefix(file.Name(), executableName) {
+				continue
+			}
+			if strings.Contains(file.Name(), ".log.INFO.") {
+				infoLogFiles = append(infoLogFiles, file)
+			} else if strings.Contains(file.Name(), ".log.WARNING.") {
+				warningLogFiles = append(warningLogFiles, file)
+			} else if strings.Contains(file.Name(), ".log.ERROR.") {
+			} else {
+				continue
+			}
+			assert.True(t, file.Size() <= 1*1024*1024, "log file size check: "+file.Name())
+		}
+		infoLogFileNum = len(infoLogFiles)
+		warningLogFileNum = len(warningLogFiles)
+	}
+
+	validateFn()
+	assert.True(t, infoLogFileNum > 2, "info log file number before checking")
+	assert.True(t, warningLogFileNum > 2, "warning log file number before checking")
+	// Call checkLogFiles() to delete extra files.
+	checkLogFiles()
+	validateFn()
+	assert.Equal(t, 2, infoLogFileNum, "info log file number after checking")
+	assert.Equal(t, 2, warningLogFileNum, "warning log file number after checking")
+
+	restoreFlagDefaultValues()
```

Comment:
Not sure whether there are other exits in the function, but is `defer restoreFlagDefaultValues()` more safe and go?

Reply:
Changed to use defer.